# Contents Based Filtering_tmdb5000

## 영화정보 전처리

In [2]:
import numpy as np
import pandas as pd
import warnings; warnings.filterwarnings('ignore')

movies = pd.read_csv('tmdb_5000_movies.csv')
movies

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4798     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4799       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4800          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4801          0                                                 []   
4802          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4798                                                NaN    9367   
4799                                                NaN   72766   
4800  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4801                        http://shanghaicalling.com/  126186   
4802                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4798  [{"id": 5616, "name": "united states\u2013mexi...                es   
4799                                                 []                en   
4800  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4801                                                 []                en   
4802  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4798                               El Mariachi   
4799                                 Newlyweds   
4800                 Signed, Sealed, Delivered   
4801                          Shanghai Calling   
4802                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4798  El Mariachi just wants to play his guitar and ...   14.269792   
4799  A newlywed couple's honeymoon is upended by th...    0.642552   
4800  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4801  When ambitious New York attorney Sam is sent t...    0.857008   
4802  Ever since the second grade when he first saw ... 

In [3]:
movies_df = movies[['id','title','genres','vote_average','vote_count','popularity',
'keywords','overview']]
movies_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4798    9367                               El Mariachi   
4799   72766                                 Newlyweds   
4800  231617                 Signed, Sealed, Delivered   
4801  126186                          Shanghai Calling   
4802   25975                         My Date with Drew   

                                                 genres  vote_average  \
0     [{"id": 28, "name": "Action"}, {"id": 12, "nam...           7.2   
1     [{"id": 12, "name": "Adventure"}, {"id": 14, "...           6.9   
2     [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.3   
3     [{"id": 28, "name": "Action"}, {"id": 80, "nam...           7.6   
4     [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.1   
...                                                 ...           ...   
4798  [{"id": 28, "name": "Action"}, {"id": 80, "nam...           6.6   
4799  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...           5.9   
4800  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...           7.0   
4801                                                 []           5.7   
4802                [{"id": 99, "name": "Documentary"}]           6.3   

      vote_count  popularity  \
0          11800  150.437577   
1           4500  139.082615   
2           4466  107.376788   
3           9106  112.312950   
4           2124   43.926995   
...          ...         ...   
4798         238   14.269792   
4799           5    0.642552   
4800           6    1.444476   
4801           7    0.857008   
4802          16    1.929883   

                                               keywords  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4     [{"id": 818, "name": "based on novel"}, {"id":...   
...                                                 ...   
4798  [{"id": 5616, "name": "united states\u2013mexi...   
4799                                                 []   
4800  [{"id": 248, "name": "date"}, {"id": 699, "nam...   
4801                                                 []   
4802  [{"id": 1523, "name": "obsession"}, {"id": 224...   

                                               overview  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Following the death of District Attorney Harve...  
4     John Carter is a war-weary, former military ca...  
...                                                 ...  
4798  El Mariachi just wants to play his guitar and ...  
4799  A newlywed couple's honeymoon is upended by th...  
4800  "Signed, Sealed, Delivered" introduces a dedic...  
4801  When ambitious New York attorney Sam is sent t...  
4802  Ever since the second grade when he first saw ...  

[4803 rows x 8 columns]

In [4]:
pd.set_option('max_colwidth',100) # 한열당 100자까지 보도록 설정
movies_df[['keywords','genres']] 

keywords  \
0     [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "name": "drug abuse"}, {"id": 911, "name": "exotic is...   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name": "based on novel"}, {"id": 4289, "name": "secret...   
3     [{"id": 849, "name": "dc comics"}, {"id": 853, "name": "crime fighter"}, {"id": 949, "name": "te...   
4     [{"id": 818, "name": "based on novel"}, {"id": 839, "name": "mars"}, {"id": 1456, "name": "medal...   
...                                                                                                   ...   
4798  [{"id": 5616, "name": "united states\u2013mexico barrier"}, {"id": 33649, "name": "legs"}, {"id"...   
4799                                                                                                   []   
4800  [{"id": 248, "name": "date"}, {"id": 699, "name": "love at first sight"}, {"id": 2398, "name": "...   
4801                                                                                                   []   
4802  [{"id": 1523, "name": "obsession"}, {"id": 2249, "name": "camcorder"}, {"id": 9986, "name": "cru...   

                                                                                                   genres  
0     [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...  
1          [{"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 28, "name": "Action"}]  
2            [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 80, "name": "Crime"}]  
3     [{"id": 28, "name": "Action"}, {"id": 80, "name": "Crime"}, {"id": 18, "name": "Drama"}, {"id": ...  
4     [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 878, "name": "Science Fic...  
...                                                                                                   ...  
4798          [{"id": 28, "name": "Action"}, {"id": 80, "name": "Crime"}, {"id": 53, "name": "Thriller"}]  
4799                                     [{"id": 35, "name": "Comedy"}, {"id": 10749, "name": "Romance"}]  
4800  [{"id": 35, "name": "Comedy"}, {"id": 18, "name": "Drama"}, {"id": 10749, "name": "Romance"}, {"...  
4801                                                                                                   []  
4802                                                                  [{"id": 99, "name": "Documentary"}]  

[4803 rows x 2 columns]

In [5]:
## 매우 중요 :  ast 모듈의 liter_eval()을 사용하면 문자열을 텍스트가 아닌 list[dict1,dict2]객체를 부를 수 있음
from ast import literal_eval
movies_df['genres'] = movies_df['genres'].apply(literal_eval)
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)

# 단순한 리스트 형태로 전처리
movies_df['genres'] = movies_df['genres'].apply(lambda x : [y['name'] for y in x])
movies_df['keywords'] = movies_df['keywords'].apply(lambda x : [y['name'] for y in x])

movies_df[['keywords','genres']]

keywords  \
0     [culture clash, future, space war, space colony, society, space travel, futuristic, romance, spa...   
1     [ocean, drug abuse, exotic island, east india trading company, love of one's life, traitor, ship...   
2                [spy, based on novel, secret agent, sequel, mi6, british secret service, united kingdom]   
3     [dc comics, crime fighter, terrorist, secret identity, burglar, hostage drama, time bomb, gotham...   
4     [based on novel, mars, medallion, space travel, princess, alien, steampunk, martian, escape, edg...   
...                                                                                                   ...   
4798                                 [united states–mexico barrier, legs, arms, paper knife, guitar case]   
4799                                                                                                   []   
4800                           [date, love at first sight, narration, investigation, team, postal worker]   
4801                                                                                                   []   
4802                                                            [obsession, camcorder, crush, dream girl]   

                                             genres  
0     [Action, Adventure, Fantasy, Science Fiction]  
1                      [Adventure, Fantasy, Action]  
2                        [Action, Adventure, Crime]  
3                  [Action, Crime, Drama, Thriller]  
4              [Action, Adventure, Science Fiction]  
...                                             ...  
4798                      [Action, Crime, Thriller]  
4799                              [Comedy, Romance]  
4800             [Comedy, Drama, Romance, TV Movie]  
4801                                             []  
4802                                  [Documentary]  

[4803 rows x 2 columns]

## 장르콘텐츠 유사도 측정

In [11]:
from sklearn.feature_extraction.text import CountVectorizer

# CountVectorizer 적용하기 위해 공백문자로 word단위가 구분되는 문자열 변환

movies_df['genres_literal'] = movies_df['genres'].apply(lambda x : (' ').join(x))
count_vect = CountVectorizer(min_df=0, ngram_range=(1,2))
genre_mat = count_vect.fit_transform(movies_df['genres_literal'])
genre_mat
# dir(genre_mat)

<4803x276 sparse matrix of type '<class 'numpy.int64'>'
	with 20631 stored elements in Compressed Sparse Row format>

276개의 장르 피처를 지닌 4803행의 매트릭스 생성 <br>
이제 cosine유사도로 거리를 측정한다

In [15]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
genre_sim

array([[1.        , 0.59628479, 0.4472136 , ..., 0.        , 0.        ,
        0.        ],
       [0.59628479, 1.        , 0.4       , ..., 0.        , 0.        ,
        0.        ],
       [0.4472136 , 0.4       , 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [16]:
# 점수는 계산 되었으니, 각 레이블 별로 높은 순위부터 장르 정렬할 필요 있음 : 넘파이의 argsort() 사용
genre_sim_sorted_idx = genre_sim.argsort()[:,::-1] # 내림차순으로 우선정렬
genre_sim_sorted_idx


array([[   0, 3494,  813, ..., 3038, 3037, 2401],
       [ 262,    1,  129, ..., 3069, 3067, 2401],
       [   2, 1740, 1542, ..., 3000, 2999, 2401],
       ...,
       [4800, 3809, 1895, ..., 2229, 2230,    0],
       [4802, 1594, 1596, ..., 3204, 3205,    0],
       [4802, 4710, 4521, ..., 3140, 3141,    0]])

## 순위 추천함수 불러오기

In [19]:

def find_sim_movies(df, sorted_idx, title_name, top_n = 10):
    # 인자로 입력된 movie_df에서 'title' 컬럼이 입력된 title_name 값인 DataFrame추출
    title_movie = df[df['title'] == title_name]

    # title_name을 가진 df의 index객체를 ndarray로 반환하고
    # sorted_idx 입력된 genre_sim_sorted_idx에서 유사도 순으로 top_n의 idx추출
    title_idx = title_movie.index.values
    similar_idx = sorted_idx[title_idx, : (top_n)]

    print(similar_idx)
    similar_idx = similar_idx.reshape(-1)
    return df.iloc[similar_idx] 

In [20]:
similar_movies = find_sim_movies(movies_df, genre_sim_sorted_idx, 'The Godfather')

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


In [21]:
similar_movies

id                                           title          genres  \
2731     240                          The Godfather: Part II  [Drama, Crime]   
1243     203                                    Mean Streets  [Drama, Crime]   
3636   36351                                   Light Sleeper  [Drama, Crime]   
1946   11699  The Bad Lieutenant: Port of Call - New Orleans  [Drama, Crime]   
2640     400         Things to Do in Denver When You're Dead  [Drama, Crime]   
4065  364083                                      Mi America  [Drama, Crime]   
1847     769                                      GoodFellas  [Drama, Crime]   
4217    9344                                            Kids  [Drama, Crime]   
883      640                             Catch Me If You Can  [Drama, Crime]   
3866     598                                     City of God  [Drama, Crime]   

      vote_average  vote_count  popularity  \
2731           8.3        3338  105.792936   
1243           7.2         345   17.002096   
3636           5.7          15    6.063868   
1946           6.0         326   17.339852   
2640           6.7          85    6.932221   
4065           0.0           0    0.039007   
1847           8.2        3128   63.654244   
4217           6.8         279   13.291991   
883            7.7        3795   73.944049   
3866           8.1        1814   44.356711   

                                                                                                 keywords  \
2731  [italo-american, cuba, vororte, melancholy, praise, revenge, mafia, lawyer, blood, corrupt polit...   
1243                                         [epilepsy, protection money, secret love, money, redemption]   
3636                                           [suicide, drug dealer, redemption, addict, existentialism]   
1946     [police brutality, organized crime, policeman, illegal drugs, murder investigation, corrupt cop]   
2640  [father son relationship, bounty hunter, boat, way of life, coffin, denver, godmother, paranoia,...   
4065                                                                         [new york state, hate crime]   
1847  [prison, based on novel, florida, 1970s, mass murder, irish-american, drug traffic, biography, b...   
4217                                                                                [puberty, first time]   
883   [con man, biography, fbi agent, overhead camera shot, attempted jailbreak, engagement party, mis...   
3866  [male nudity, street gang, brazilian, photographer, 1970s, puberty, ghetto, gang war, coming of ...   

                                                                                                 overview  \
2731  In the continuing saga of the Corleone crime family, a young Vito Corleone grows up in Sicily an...   
1243     A small-time hood must choose from among love, friendship and the chance to rise within the mob.   
3636  A drug dealer with upscale clientele is having moral problems going about his daily deliveries. ...   
1946  Terrence McDonagh, a New Orleans Police sergeant, who starts out as a good cop, receiving a meda...   
2640  A mafia film in Tarantino style with a star-studded cast. Jimmy’s “The Saint” gangster career ha...   
4065  A hate-crime has been committed in a the small city of Braxton, N.Y. Five migrant laborers have ...   
1847  The true story of Henry Hill, a half-Irish, half-Sicilian Brooklyn kid who is adopted by neighbo...   
4217  A controversial portrayal of teens in New York City which exposes a deeply disturbing world of s...   
883   A true story about Frank Abagnale Jr. who, before his 19th birthday, successfully conned million...   
3866  Cidade de Deus is a shantytown that started during the 1960s and became one of Rio de Janeiro’s ...   

     genres_literal  
2731    Drama Crime  
1243    Drama Crime  
3636    Drama Crime  
1946    Drama Crime  
2640    Drama Crime  
4065    Drama Crime  
1847    Drama Crime  
4217    Drama Crime  
883     Drama Crime  
38

듣보잡 영화도 장르만으로 가져오다보니, 많이 끌고 온 것 확인. 어쩌지? <br>
가중치가 부여된 평점을 활용해서 가중치를 주고자 함 <br>

- 개별영화 평점투표 횟수 : v
- 평점부여 최소투표 횟수 : m (투표횟수에 직접 가중치를 조정하는 역할을 함, m을 높이면 평점투표횟수가 많은 영화에 더 많은 가중평점 부여)
- 개별영화 평균평점 : R
- 전체영화 평균평점 : C
<br><br>
- 가중평점 = (v/(v+m)) * R + (m/(v+m)) * C



In [22]:
C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.6) # 전체투표중에서 상위 60%에 해당하는 횟수를 기준으로 정함

def weighted_vote_average(record):
    v = record['vote_count']
    R = record['vote_average']

    return ((v/(v+m)) * R) + ((m/(v+m)) * C)

movies_df['weighted_vote'] = movies_df.apply(weighted_vote_average, axis=1)

In [24]:
movies_df.sort_values('weighted_vote', ascending=False)[:10] # weighted_vote 상위 10개 영화 확인

id                     title  \
1881     278  The Shawshank Redemption   
3337     238             The Godfather   
662      550                Fight Club   
3232     680              Pulp Fiction   
65       155           The Dark Knight   
1818     424          Schindler's List   
3865  244786                  Whiplash   
809       13              Forrest Gump   
2294     129             Spirited Away   
2731     240    The Godfather: Part II   

                                       genres  vote_average  vote_count  \
1881                           [Drama, Crime]           8.5        8205   
3337                           [Drama, Crime]           8.4        5893   
662                                   [Drama]           8.3        9413   
3232                        [Thriller, Crime]           8.3        8428   
65           [Drama, Action, Crime, Thriller]           8.2       12002   
1818                    [Drama, History, War]           8.3        4329   
3865                                  [Drama]           8.3        4254   
809                  [Comedy, Drama, Romance]           8.2        7927   
2294  [Fantasy, Adventure, Animation, Family]           8.3        3840   
2731                           [Drama, Crime]           8.3        3338   

      popularity  \
1881  136.747729   
3337  143.659698   
662   146.757391   
3232  121.463076   
65    187.322927   
1818  104.469351   
3865  192.528841   
809   138.133331   
2294  118.968562   
2731  105.792936   

                                                                                                 keywords  \
1881  [prison, corruption, police brutality, prison cell, delinquent, parole board, escape from prison...   
3337  [italy, love at first sight, loss of father, patriarch, organized crime, mafia, lawyer, italian ...   
662                 [support group, dual identity, nihilism, rage and hate, insomnia, dystopia, violence]   
3232  [transporter, brothel, drug dealer, boxer, massage, stolen money, crime boss, dance contest, jun...   
65    [dc comics, crime fighter, secret identity, scarecrow, sadism, chaos, gotham city, vigilante, jo...   
1818  [factory, concentration camp, hero, holocaust, world war ii, nazis, defense industry, biography,...   
3865  [jazz, obsession, conservatory, music teacher, new york city, violence, montage, drummer, public...   
809   [vietnam veteran, hippie, mentally disabled, running, based on novel, vietnam, vietnam war, frie...   
2294  [witch, parents kids relationship, magic, twilight, darkness, village and town, bath house, pig,...   
2731  [italo-american, cuba, vororte, melancholy, praise, revenge, mafia, lawyer, blood, corrupt polit...   

                                                                                                 overview  \
1881  Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufr...   
3337  Spanning the years 1945 to 1955, a chronicle of the fictional Italian-American Corleone crime fa...   
662   A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a...   
3232  A burger-loving hit man, his philosophical partner, a drug-addled gangster's moll and a washed-u...   
65    Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attor...   
1818  The true story of how businessman Oskar Schindler saved over a thousand Jewish lives from the Na...   
3865  Under the direction of a ruthless instructor, a talented young drummer begins to pursue perfecti...   
809   A man with a low IQ has accomplished great things in his life and been present during significan...   
2294  A ten year old girl who wanders away from her parents along a path that leads to a world ruled b...   
2731  In the continuing saga of the Corleone crime family, a young Vito Corleone grows up in Sicily an...   

                          genres_literal  weighted_vote  
1881                         Drama Crime  

In [25]:

def find_sim_movies_weighted(df, sorted_idx, title_name, top_n = 10):
    # 인자로 입력된 movie_df에서 'title' 컬럼이 입력된 title_name 값인 DataFrame추출
    title_movie = df[df['title'] == title_name]

    # title_name을 가진 df의 index객체를 ndarray로 반환하고
    # sorted_idx 입력된 genre_sim_sorted_idx에서 유사도 순으로 top_n의 idx추출
    title_idx = title_movie.index.values
    similar_idx = sorted_idx[title_idx, : (top_n*2)] # 두배 더 뽑고
    similar_idx = similar_idx.reshape(-1)
    similar_idx = similar_idx[similar_idx != title_idx] # 기준영화명은 제외

    return df.iloc[similar_idx].sort_values('weighted_vote', ascending = False)[:top_n] # weighted_vote로 재정렬

In [26]:
similar_movies_weighted = find_sim_movies_weighted(movies_df, genre_sim_sorted_idx, 'The Godfather')
similar_movies_weighted

id                        title          genres  vote_average  \
2731     240       The Godfather: Part II  [Drama, Crime]           8.3   
1847     769                   GoodFellas  [Drama, Crime]           8.2   
3866     598                  City of God  [Drama, Crime]           8.1   
1663     311  Once Upon a Time in America  [Drama, Crime]           8.2   
883      640          Catch Me If You Can  [Drama, Crime]           7.7   
281     4982            American Gangster  [Drama, Crime]           7.4   
4041   11798              This Is England  [Drama, Crime]           7.4   
1149  168672              American Hustle  [Drama, Crime]           6.8   
1243     203                 Mean Streets  [Drama, Crime]           7.2   
2839   10220                     Rounders  [Drama, Crime]           6.9   

      vote_count  popularity  \
2731        3338  105.792936   
1847        3128   63.654244   
3866        1814   44.356711   
1663        1069   49.336397   
883         3795   73.944049   
281         1502   42.361215   
4041         363    8.395624   
1149        2807   49.664128   
1243         345   17.002096   
2839         439   18.422008   

                                                                                                 keywords  \
2731  [italo-american, cuba, vororte, melancholy, praise, revenge, mafia, lawyer, blood, corrupt polit...   
1847  [prison, based on novel, florida, 1970s, mass murder, irish-american, drug traffic, biography, b...   
3866  [male nudity, street gang, brazilian, photographer, 1970s, puberty, ghetto, gang war, coming of ...   
1663  [life and death, corruption, street gang, rape, sadistic, lovesickness, sexual abuse, money laun...   
883   [con man, biography, fbi agent, overhead camera shot, attempted jailbreak, engagement party, mis...   
281   [underdog, black people, drug traffic, drug smuggle, society, ambition, rise and fall, cop, drug...   
4041  [holiday, skinhead, england, vandalism, independent film, gang, racism, summer, youth, violence,...   
1149                                                               [con artist, scam, mobster, fbi agent]   
1243                                         [epilepsy, protection money, secret love, money, redemption]   
2839                                                 [gambling, law, compulsive gambling, roulette, gain]   

                                                                                                 overview  \
2731  In the continuing saga of the Corleone crime family, a young Vito Corleone grows up in Sicily an...   
1847  The true story of Henry Hill, a half-Irish, half-Sicilian Brooklyn kid who is adopted by neighbo...   
3866  Cidade de Deus is a shantytown that started during the 1960s and became one of Rio de Janeiro’s ...   
1663  A former Prohibition-era Jewish gangster returns to the Lower East Side of Manhattan over thirty...   
883   A true story about Frank Abagnale Jr. who, before his 19th birthday, successfully conned million...   
281   Following the death of his employer and mentor, Bumpy Johnson, Frank Lucas establishes himself a...   
4041  A story about a troubled boy growing up in England, set in 1983. He comes across a few skinheads...   
1149  A con man, Irving Rosenfeld, along with his seductive partner Sydney Prosser, is forced to work ...   
1243     A small-time hood must choose from among love, friendship and the chance to rise within the mob.   
2839  A young man is a reformed gambler who must return to playing big stakes poker to help a friend p...   

     genres_literal  weighted_vote  
2731    Drama Crime       8.079586  
1847    Drama Crime       7.976937  
3866    Drama Crime       7.759693  
1663    Drama Crime       7.657811  
883     Drama Crime       7.557097  
281     Drama Crime       7.141396  
4041    Drama Crime       6.739664  
1149    Drama Crime       6.717525  
1243    Drama Crime       6.626569  
2839    Drama Crime       6.530427

인기 높은 영화를 고루 추천하는 것 확인